crypto/context.c

```diff
@@ -14,27 +14,39 @@
 #include "internal/core.h"
 #include "internal/bio.h"
 #include "internal/provider.h"
+#include "crypto/context.h"
 
 struct ossl_lib_ctx_onfree_list_st {
     ossl_lib_ctx_onfree_fn *fn;
     struct ossl_lib_ctx_onfree_list_st *next;
 };
 
 struct ossl_lib_ctx_st {
-    CRYPTO_RWLOCK *lock;
-    CRYPTO_EX_DATA data;
-
-    /*
-     * For most data in the OSSL_LIB_CTX we just use ex_data to store it. But
-     * that doesn't work for ex_data itself - so we store that directly.
-     */
+    CRYPTO_RWLOCK *lock, *rand_crngt_lock;
     OSSL_EX_DATA_GLOBAL global;
 
-    /* Map internal static indexes to dynamically created indexes */
-    int dyn_indexes[OSSL_LIB_CTX_MAX_INDEXES];
-
-    /* Keep a separate lock for each index */
-    CRYPTO_RWLOCK *index_locks[OSSL_LIB_CTX_MAX_INDEXES];
+    void *property_string_data;
+    void *evp_method_store;
+    void *provider_store;
+    void *namemap;
+    void *property_defns;
+    void *global_properties;
+    void *drbg;
+    void *drbg_nonce;
```

Comment:
Not for this PR but it would be really nice if these fields had a type which the compiler could check....

I know that means not using the one accessor function etc.


We have altogether too many void pointers being cast.  Everything across the provider boundary is a void * and mistakes have been made there.